hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -1481,18 +1487,21 @@ protected HoodieData<HoodieRecord> prepRecords(Map<String, HoodieData<HoodieReco
         ValidationUtils.checkArgument(fileGroupCount > 0, String.format("FileGroup count for MDT partition %s should be > 0", partitionName));
 
         List<FileSlice> finalFileSlices = fileSlices;
+        Set<String> mappedFileIds = new HashSet<>();
         HoodieData<HoodieRecord> rddSinglePartitionRecords = records.map(r -> {
           FileSlice slice = finalFileSlices.get(HoodieTableMetadataUtil.mapRecordKeyToFileGroupIndex(r.getRecordKey(),
               fileGroupCount));
           r.unseal();
           r.setCurrentLocation(new HoodieRecordLocation(slice.getBaseInstantTime(), slice.getFileId()));
           r.seal();
+          mappedFileIds.add(slice.getFileId());
```

Comment:
The `mappedFileIds` set will be serialized to the executor in the spark context. It will not be reported back to the driver so I don't think this implementation will work.

Reply:
good catch. I did fix it in my later patch, but missed to update here. 

https://github.com/apache/hudi/pull/13312/files#r2112561019 
you can check it out here. 

there is a known limitation to this. we will be spinning up a spark task while writing for all file groups for the mdt partitions we are touching, since w/o triggering the action on the records, we don't really know which among the file groups actually have records to be written to. 

but since we do not want to trigger any action, we are returning every file group in the partiions we touch from here.

Reply:
will fix this patch accordingly.